pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go

```diff
@@ -483,57 +388,40 @@ func (w *Webhook) getAutoDetectedLibraries(pod *corev1.Pod) []libInfo {
 }
 
 func (w *Webhook) extractLibrariesFromAnnotations(pod *corev1.Pod) []libInfo {
-	annotations := pod.Annotations
-	var libList []libInfo
-	for _, lang := range supportedLanguages {
-		customLibAnnotation := strings.ToLower(fmt.Sprintf(customLibAnnotationKeyFormat, lang))
-		if image, found := annotations[customLibAnnotation]; found {
-			log.Debugf(
-				"Found %s library annotation %s, will overwrite %s injected with Single Step Instrumentation",
-				string(lang), customLibAnnotation, image,
-			)
-			libList = append(libList, libInfo{lang: lang, image: image})
-		}
-
-		libVersionAnnotation := strings.ToLower(fmt.Sprintf(libVersionAnnotationKeyFormat, lang))
-		if version, found := annotations[libVersionAnnotation]; found {
-			image := fmt.Sprintf("%s/dd-lib-%s-init:%s", w.containerRegistry, lang, version)
-			log.Debugf(
-				"Found %s library annotation for version %s, will overwrite %s injected with Single Step Instrumentation",
-				string(lang), version, image,
-			)
-			libList = append(libList, libInfo{lang: lang, image: image})
+	var (
+		libList []libInfo
+		extract = func(e annotationExtractor[libInfo]) {
+			if i, ok, _ := e.extract(pod); ok {
+				libList = append(libList, i)
+			}
 		}
+	)
 
+	for _, l := range supportedLanguages {
+		extract(l.customLibAnnotationExtractor())
+		extract(l.libVersionAnnotationExtractor(w.containerRegistry))
 		for _, ctr := range pod.Spec.Containers {
-			customLibAnnotation := strings.ToLower(fmt.Sprintf(customLibAnnotationKeyCtrFormat, ctr.Name, lang))
-			if image, found := annotations[customLibAnnotation]; found {
-				log.Debugf(
-					"Found custom library annotation for %s, will inject %s to container %s",
-					string(lang), image, ctr.Name,
-				)
-				libList = append(libList, libInfo{ctrName: ctr.Name, lang: lang, image: image})
-			}
-
-			libVersionAnnotation := strings.ToLower(fmt.Sprintf(libVersionAnnotationKeyCtrFormat, ctr.Name, lang))
-			if version, found := annotations[libVersionAnnotation]; found {
-				image := libImageName(w.containerRegistry, lang, version)
-				log.Debugf(
-					"Found version library annotation for %s, will inject %s to container %s",
-					string(lang), image, ctr.Name,
-				)
-				libList = append(libList, libInfo{ctrName: ctr.Name, lang: lang, image: image})
-			}
+			extract(l.ctrCustomLibAnnotationExtractor(ctr.Name))
+			extract(l.ctrLibVersionAnnotationExtractor(ctr.Name, w.containerRegistry))
```

Comment:
Example of using the `annotationExtractor[libInfo]` instead of this logic is the simplification to get us something that's a bit clearer in what it's doing.